xapian-applications/omega/query.cc

```diff
@@ -1444,7 +1452,11 @@ eval(const string &fmt, const vector<string> &param)
 		value = args[0];
 		if (!value.empty()) {
 		    char buf[64] = "";
-		    time_t date = string_to_int(value);
+		    unsigned int temp;
+		    if (!parse_unsigned(value.c_str(), temp)) {
+			throw "Date(in secs) for command date should be >= 0";
+		    }
```

Comment:
Dates before 1970 have negative `time_t` values (and currently work).

I wonder if we want a `parse_signed()` function and to use that to read directly into a `time_t` (which POSIX requires to be a signed integer type IIRC).  Using `int` here means that dates after somewhere in 2038 don't work.

Reply:
> I wonder if we want a `parse_signed()` function 
>
yeah that would be a good idea, It would lead to a much clearer and less code too.  I will write one for that. I think long long should solve the date after 2038 problem.

> Dates before 1970 have negative `time_t` values (and currently work).
> 
Didn't know that, will fix it.

Reply:
> I think long long should solve the date after 2038 problem.

I don't think using a wider type really helps (and if you wanted a 64-bit, use `int64_t` - `long long` is only guaranteed to be **at least** a 64 bit type - it could be wider).

But if you parse the value into `long long`, then you'll have to decide what to do if the parsed integer value doesn't fit in a `time_t` - if you just assign it without checking then it's likely to just mask off the upper bits and dates after the highest representable time in 2038 suddenly come out as in 1901 which is just not helpful.

We could throw an exception, but then searching the same database can throw an exception on a platform with 32-bit `time_t` when it works on a platform with 64-bit `time_t`.  But if you wanted to do that, then trying to convert to `time_t` directly seems the best way to do it.

Probably the best alternative is for `$date{X}` to return an empty string if the date can't be represented on the current platform.  But then it seems more consistent to do that for **any** value which doesn't fit in a `time_t`, so converting via `int64_t` or `long long` just makes things more complicated - the proposed `parse_signed()` will tell us if the value doesn't fit in a `time_t`, and then we don't try to convert it further.

Reply:
understood.